detox/src/devices/Device.js

```diff
@@ -180,6 +181,19 @@ class Device {
     await this.deviceDriver.uninstallApp(this._deviceId, _bundleId);
   }
 
+  async installUtilBinaries() {
+    const paths = this._deviceConfig.utilBinaryPaths;
+    if (paths) {
+      if (!_.isArray(paths)) {
+        throw new DetoxRuntimeError({
+          message: `utilBinaryPaths specified in Detox configuration is invalid: ${paths}`,
+          hint: 'Configuration must be applied as an Array of paths',
+        });
+      }
+      await this.deviceDriver.installUtilBinaries(this._deviceId, paths);
```

Comment:
Can't we make 1-size-fits-all `this.deviceDriver.installApp(deviceId, binaryPath)` signature?

Is there any fundemantal difference between `this.deviceDriver.installUtilBinaries` and `this.deviceDriver.installApp` except for spread arguments?

Reply:
I took the time to think about it and decided not to because the use-case and impl. here are different, and because I like being explicit (it makes it easier to make changes, later...). To prove my point, `installApp` currently expects two binaries and as a contract uses a default if no test-binary is specified, and I don't want to break it.

Reply:
@d4vidi , let's check again we are on the same page:

> 1. To prove my point, installApp currently expects two binaries and as a contract uses a default if no test-binary is specified.

First of all, isn't it a bad contract, considering that iOS simply does not need this quirk at all? I reckon that the optional `testBinary` parameter is not generic, rather quite platform- and implementation-specific, no?

On the other hand, I have considered the difficulty of moving these parts, and currently, I don't see a cleaner solution than the one you had suggested. Although, I feel like a heavy technical debt is forming here. Hence, I want to ask you in a roadmap context (and not of this PR):

Don't you think that a wiser decision would be to reimplement eventually `Device` class by splitting it into several stateful platform-specific implementations such as `SimulatorDevice`, `AttachedAndroidDevice`, `AndroidEmulatorDevice` and so on? The fact that we have a monolithic one-size-fits-all `Device` that has a complex state, yet it manages (with a struggle, albeit) to somehow function by juggling stateless `*DeviceDriver` methods has been saddening me for quite long. Making Device subclasses does not cancel code reuse, anyway, but the current way of handling is making configuration messy while making superficial visibility that is uniform.

Won't it be just relieving to have
```
{ type: 'ios.simulator' } -> gets handled by SimulatorDevice, that has its own schema
{ type: 'android.emulator' } -> gets handled by EmulatorDevice, that has its own and another schema
```
?

Reply:
Honestly, it's like you've read my mind! I went back and forth with this idea in my head a fair share of times.
I definitely agree the the state-less-ness of the drivers has no value. Seriously, passing in the `deviceId` in each call? this is a high-level coding environment... we should apply more advanced techniques (sorry for mocking).

As per the lack of their necessity altogether? I can't clearly conclude I agree with _that_. There's value in delegating the platform-specifics onto a different domain, making the `Device` still inheritable potentially for other (more profitable?) needs. But being a bloated class that it is right now? probably a bad idea.